ethcore/src/miner/zero_gas_price_checker.rs

```diff
@@ -0,0 +1,237 @@
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity. If not, see <http://www.gnu.org/licenses/>.
+
+use client::MiningBlockChainClient;
+use transaction::SignedTransaction;
+use util::{U256, Uint};
+
+const ZGP_CONTRACT_REGISTRY_NAME: &'static str = "zgp_checker";
+
+/// Zero gas price transactions checker.
+pub trait ZeroGasPriceChecker: Send {
+	/// Checks if zgp-transaction can be appended to the transaction queue.
+	fn check(&self, chain: Option<&MiningBlockChainClient>, tx: &SignedTransaction) -> Result<bool, String>;
+}
+
+/// Implementation of `ZeroGasPriceChecker`, based on on-chain deployed contract.
+pub struct OnChainZeroGasPriceChecker {
+	contract: provider::Contract,
+}
+
+impl OnChainZeroGasPriceChecker {
+	/// Try to create instance, reading contract address from given chain.
+	pub fn from_chain(chain: &MiningBlockChainClient) -> Option<Self> {
+		chain.registry_address(ZGP_CONTRACT_REGISTRY_NAME.to_owned())
+			.and_then(|contract_addr| {
+				trace!(target: "txqueue", "Configuring for zgp-contract from {}", contract_addr);
+
+				Some(OnChainZeroGasPriceChecker {
+					contract: provider::Contract::new(contract_addr),
+				})
+			})
+	}
+}
+
+impl ZeroGasPriceChecker for OnChainZeroGasPriceChecker {
+	/// Check if we can append zgp-transaction to the transaction queue
+	fn check(&self, chain: Option<&MiningBlockChainClient>, tx: &SignedTransaction) -> Result<bool, String> {
+		debug_assert_eq!(tx.gas_price, U256::zero());
+
+		chain.ok_or("Client not set".to_owned())
+			.and_then(|chain| {
+				let do_call = |a, d| chain.call_contract(a, d);
+				self.contract.certified(&do_call, &tx.sender())
+			})
+	}
+}
+
+#[cfg(test)]
+pub mod tests {
+	use std::collections::HashMap;
+	use client::MiningBlockChainClient;
+	use transaction::SignedTransaction;
+	use util::H256;
+	use super::ZeroGasPriceChecker;
+
+	#[derive(Debug, Default)]
+	pub struct DummyZeroGasPriceChecker {
+		pub check_results: HashMap<H256, Result<bool, String>>,
+	}
+
+	impl ZeroGasPriceChecker for DummyZeroGasPriceChecker {
+		fn check(&self, _chain: Option<&MiningBlockChainClient>, tx: &SignedTransaction) -> Result<bool, String> {
+			self.check_results.get(&tx.hash()).cloned().unwrap_or(Ok(true))
+		}
+	}
+}
+
+mod provider {
+	// Autogenerated from JSON contract definition using Rust contract convertor.
+	// Command line: --jsonabi=SimpleCertifier.abi --explicit-do-call
+	#![allow(unused_imports)]
+	use std::string::String;
+	use std::result::Result;
+	use std::fmt;
+	use {util, ethabi};
+	use util::{FixedHash, Uint};
+
+	pub struct Contract {
+		contract: ethabi::Contract,
+		address: util::Address,
+		
+	}
+	impl Contract {
+		pub fn new(address: util::Address) -> Self
+			{
```

Comment:
minor style grumble: `{` goes at end of previous line, not on own line (and if it were to go on own line, it would be at lower indentation level).